src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp

```diff
@@ -450,6 +451,13 @@ namespace {
     }
 
 
+    void Schedule::addACTIONX(const Action::ActionX& action, std::size_t currentStep) {
+        auto new_actions = std::make_shared<Action::Actions>( this->actions(currentStep) );
+        new_actions->add(action);
+        this->m_actions.update(currentStep, new_actions);
```

Comment:
The intention of the update() call is clear to me (I think): add another entry to the DynamicState container if the argument passed differs from that which already is in the container. But is this supposed to be a "content" comparison (i.e. comparing Actions objects) or an "identity" comparison (comparing pointers)? The implementation in DynamicState::update() makes me think an identity comparison is made when the content type is a pointer, is that the intention?

I think this could lead to identical actions being added to every timestep in spite of not being needed. If so, it is not very harmful, but eliminates the memory saving DynamicState is intended to enable.

Reply:
The `DynamicState` used value based semantics - so this: 

> The implementation in DynamicState::update() makes me think an identity comparison is made when the content type is a pointer

is correct. The "idiomatic pattern" generally used in the `Schedule`code is:
```
// Create a new something with the Something copy constructor
std::shared_ptr<Something> new_something( dynamic_state.get(step) );

// Update new_something with whatever keyword we are currently processing.
new_something->setXXX(  ... );

// Check if there has been a net change in something - deep comparison
if (*new_something != Schedule::getSomething( step ))
     dynamic_state.update( step, new_something)
```
in this particular case the deep comparison is skipped - under the assumption that *if* `ACTIONX`keyword is indeed presented in the deck the assumption is that it *will* represent a net change.

Reply:
> but eliminates the memory saving DynamicState is intended to enable.

The main space saving with the `std::shared_ptr`pattern is that if an object is updated at report steps 0 and 100 there will only be two objects stored; where all elements 0..99 will point to the same underlying object (that shared_ptr will have refcount 100).